// pertain/build/module/PackageJson.js
import dotProp from 'dot-prop';
import path from 'path';
/**
 * Lazy-loading proxy for package.json. A subset of properties are available,
 * and a static method
 * `PackageJson.lookup(package: PackageJson, dotPath: string)`
 * exists for looking up other arbitrary properties.
 *
 * This speeds up dependency gathering; it won't be necessary to read all
 * package files from disk.
 */
export default class PackageJson {
    constructor(modulePath) {
        this.modulePath = modulePath;
    }
    /**
     * Look up a custom dot-path on a package.json instance. Abstracting this
     * simple operation behind a static method helps us to keep this loading lazy.
     */
    static lookup(pkg, dotPath) {
        return dotProp.get(pkg.getJson(), dotPath);
    }
    get dependencies() {
        return this.getJson().dependencies || {};
    }
    get devDependencies() {
        return this.getJson().devDependencies || {};
    }
    get name() {
        return this.getJson().name;
    }
    get peerDependencies() {
        return this.getJson().peerDependencies || {};
    }
    getJson() {
        if (!this.json) {
            this.json = require(path.join(this.modulePath, 'package.json'));
        }
        return this.json;
    }
}
